File: day04/day4_1.js
fs = require('fs')

fs.readFile('./input.txt', 'utf8', function (err, data) {
  if (err) {
    return console.log(err);
  }

  let sum = data.split('\n').reduce(
    (acc, line) => acc + getLineValidity(line),
    0
  );

  console.log(sum);
});

function getLineValidity(line) {
  let words = line.split(' ');
  let uniques = unique(words);

  return words.length === uniques.length ? 1 : 0;
}

function unique(words) {
  let seen = {};
  return words.filter((w) => {
    if(w in seen) {
      return false;
    }
    seen[w] = undefined;
    return true;
  });
}

File: day04/day4_1.clj
(defn uniqueWords? [line]
  (let [words (clojure.string/split line #"\s")
        uniques (set words)]
    (= (count words) (count uniques))))

(count
 (with-open [rdr (clojure.java.io/reader "./input.txt")]
     (->> (line-seq rdr)
          (filter #(uniqueWords? %))
          (doall))))
